botocore/configprovider.py

```diff
@@ -0,0 +1,418 @@
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""This module contains the inteface for controlling how configuration
+is loaded.
+"""
+import os
+
+
+def create_botocore_default_config_mapping(chain_builder):
+    return {
+        'profile': chain_builder.create_config_chain(
+            instance_name='profile',
+            env_var_names=['AWS_DEFAULT_PROFILE', 'AWS_PROFILE'],
+        ),
+        'region': chain_builder.create_config_chain(
+            instance_name='region',
+            env_var_names='AWS_DEFAULT_REGION',
+            config_property_name='region',
+            default=None,
+        ),
+        'data_path': chain_builder.create_config_chain(
+            instance_name='data_path',
+            env_var_names='AWS_DATA_PATH',
+            config_property_name='data_path',
+            default=None,
+        ),
+        'config_file': chain_builder.create_config_chain(
+            instance_name='config_file',
+            env_var_names='AWS_CONFIG_FILE',
+            default='~/.aws/config',
+        ),
+        'ca_bundle': chain_builder.create_config_chain(
+            instance_name='ca_bundle',
+            env_var_names='AWS_CA_BUNDLE',
+            config_property_name='ca_bundle',
+        ),
+        'api_versions': chain_builder.create_config_chain(
+            instance_name='api_versions',
+            config_property_name='api_versions',
+            default={},
+        ),
+        'credentials_file': chain_builder.create_config_chain(
+            instance_name='credentials_file',
+            env_var_names='AWS_SHARED_CREDENTIALS_FILE',
+            default='~/.aws/credentials',
+        ),
+        'metadata_service_timeout': chain_builder.create_config_chain(
+            instance_name='metadata_service_timeout',
+            env_var_names='AWS_METADATA_SERVICE_TIMEOUT',
+            config_property_name='metadata_service_timeout',
+            default=1,
+            conversion_func=int,
+        ),
+        'metadata_service_num_attempts': chain_builder.create_config_chain(
+            instance_name='metadata_service_num_attempts',
+            env_var_names='AWS_METADATA_SERVICE_NUM_ATTEMPTS',
+            config_property_name='metadata_service_num_attempts',
+            default=1,
+            conversion_func=int,
+        ),
+        'parameter_validation': chain_builder.create_config_chain(
+            instance_name='parameter_validation',
+            config_property_name='parameter_validation',
+            default=True,
+        ),
+    }
+
+
+class ConfigChainFactory(object):
```

Comment:
Would you consider this part of the public API?  Seems like we really only want the `ConfigValueStore` to be considered public right?  It's worth documenting what we consider to be public here.

Reply:
Ah I hadn't thought about that specifically. I don't think there is any reason for it to be public since its a convenience class for us. It would only be useful for people that want to make the same configuration chain as we do, which I don't think is a big use case. Mainly I was considering the `ConfigStore` as the public class since it gets exposed through the `get_component('config_store')` on the session. And thats the main way to interact with overriding config providers. From there the `BaseProvider` is the interface they would use so that is public as well.

Should we add this to the botocore docs somewhere or document it in code?